tests/xtro-sharpie/DeprecatedCheck.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Clang;
+using Clang.Ast;
+using Mono.Cecil;
+
+namespace Extrospection
+{
+	public class DeprecatedCheck : BaseVisitor
+	{
+		Dictionary<string, VersionTuple> ObjCDeprecatedItems = new Dictionary<string, VersionTuple> ();
+		Dictionary<string, VersionTuple> ObjCDeprecatedSelectors = new Dictionary<string, VersionTuple> ();
+
+		List<TypeDefinition> ManagedTypes = new List<TypeDefinition> ();
+
+		public override void End ()
+		{
+			foreach (var objcEntry in ObjCDeprecatedItems)
+				ProcessObjcEntry (objcEntry.Key, objcEntry.Value);
+
+			foreach (var objcEntry in ObjCDeprecatedSelectors)
+				ProcessObjcSelector (objcEntry.Key, objcEntry.Value);
+		}
+
+		void ProcessObjcEntry (string objcClassName, VersionTuple objcVersion)
+		{
+			TypeDefinition managedType = ManagedTypes.FirstOrDefault (x => Helpers.GetName (x) == objcClassName && x.IsPublic);
+			if (managedType != null) {					
+				var framework = Helpers.GetFramework (managedType);
+				if (framework != null)
+					ProcessItem (managedType, Helpers.GetName (managedType), objcVersion, framework);
+			}
+		}
+
+		void ProcessObjcSelector (string fullname, VersionTuple objcVersion)
+		{
+			string[] nameParts = fullname.Split (new string[] { "::" }, StringSplitOptions.None);
+
+			string objcClassName = nameParts[0];
+			string selector = nameParts[1];
+
+			TypeDefinition managedType = ManagedTypes.FirstOrDefault (x => Helpers.GetName (x) == objcClassName);
+			if (managedType != null) {
+				var framework = Helpers.GetFramework (managedType);
+				if (framework == null)
+					return;
+
+				// If the entire type is deprecated, call it good enough
+				if (HasAnyDeprecationAttribute (managedType.CustomAttributes))
```

Comment:
check `HasCustomAttributes` before the call
or
change the call to accept an `ICustomAttributeProvider` argument and do the check within `HasAnyDeprecationAttribute` (might require less changes)